protocol/x/clob/keeper/process_operations.go

```diff
@@ -248,6 +214,218 @@ func (k Keeper) PersistMatchToState(
+	orderIdToRemove.MustBeStatefulOrder()
+
+	// Order removals are always for stateful orders that must exist.
+	orderToRemove, err := k.FetchOrderFromOrderId(ctx, orderIdToRemove, make(map[types.OrderId]types.Order, 0))
+	if err != nil {
+		return err
+	}
+
+	// Statefully validate that the removal reason is valid.
+	switch removalReason := orderRemoval.RemovalReason; removalReason {
+	case types.OrderRemoval_REMOVAL_REASON_UNDERCOLLATERALIZED:
+		// For the collateralization check, use the remaining amount of the order that is resting on the book.
+		remainingAmount, hasRemainingAmount := k.MemClob.GetOrderRemainingAmount(ctx, orderToRemove)
+		if !hasRemainingAmount {
+			panic(fmt.Sprintf("addOrderToOrderbookCollateralizationCheck: order has no remaining amount %v", orderToRemove))
+		}
+
+		pendingOpenOrder := types.PendingOpenOrder{
+			RemainingQuantums: remainingAmount,
+			IsBuy:             orderToRemove.IsBuy(),
+			Subticks:          orderToRemove.GetOrderSubticks(),
+			ClobPairId:        orderToRemove.GetClobPairId(),
+		}
+
+		// Temporarily construct the subaccountOpenOrders with a single PendingOpenOrder.
+		subaccountOpenOrders := map[satypes.SubaccountId][]types.PendingOpenOrder{
+			orderIdToRemove.SubaccountId: {
+				pendingOpenOrder,
+			},
+		}
+
+		// TODO(DEC-1896): AddOrderToOrderbookCollatCheck should accept a single PendingOpenOrder as a
+		// parameter rather than the subaccountOpenOrders map.
+		_, successPerSubaccountUpdate := k.AddOrderToOrderbookCollatCheck(
+			ctx,
+			orderToRemove.GetClobPairId(),
+			subaccountOpenOrders,
+		)
+		if successPerSubaccountUpdate[orderIdToRemove.SubaccountId].IsSuccess() {
+			return errorsmod.Wrapf(
+				types.ErrInvalidOrderRemoval,
+				"Order Removal (%+v) invalid. Order passes collateralization check.",
+				*orderRemoval,
+			)
+		}
+	case types.OrderRemoval_REMOVAL_REASON_INVALID_REDUCE_ONLY:
+		if !orderToRemove.IsReduceOnly() {
+			return errorsmod.Wrapf(
+				types.ErrInvalidOrderRemoval,
+				"Order Removal (%+v) invalid. Order must be reduce only.",
+				*orderRemoval,
+			)
+		}
+
+		// The reduce-only order must increase or change the side of the position to trigger removal.
+		currentPositionSize := k.GetStatePosition(
+			ctx,
+			orderIdToRemove.SubaccountId,
+			orderToRemove.GetClobPairId(),
+		)
+		orderQuantumsToFill := orderToRemove.GetBigQuantums()
+
+		orderFillWouldIncreasePositionSize := orderQuantumsToFill.Sign() == currentPositionSize.Sign()
+
+		newPositionSize := new(big.Int).Add(currentPositionSize, orderQuantumsToFill)
+		orderChangedSide := currentPositionSize.Sign()*newPositionSize.Sign() == -1
+		if !orderFillWouldIncreasePositionSize && !orderChangedSide {
+			return errorsmod.Wrapf(
+				types.ErrInvalidOrderRemoval,
+				"Order Removal (%+v) invalid. Order fill must increase position size or change side.",
+				*orderRemoval,
+			)
+		}
+	case types.OrderRemoval_REMOVAL_REASON_POST_ONLY_WOULD_CROSS_MAKER_ORDER:
+		// TODO
```

Comment:
can you add a ticket to this and other TODOs?

Reply:
for these and other removals that can not be verified without additional information. let's add a stat to keep track of the count (ideally with the proposer address in the labels)